tscat_gui/__init__.py

```diff
@@ -51,8 +51,180 @@ def __init__(self, parent=None):
+                self.programmatic_select = True
+                self.catalogues_view.clearSelection()
+                for index in indexes:
+                    self.catalogues_view.selectionModel().select(index,
+                                                                 QtCore.QItemSelectionModel.SelectionFlag.Select)
+                self.programmatic_select = False
+            else:
+                if action not in ['active_select', 'passive_select']:
+                    self.events_model.reset()
+
+                indexes = list(map(self.events_model.index_from_uuid, uuids))
+                indexes = list(map(self.events_sort_model.mapFromSource, indexes))
+                self.programmatic_select = True
+                self.events_view.clearSelection()
+                for index in indexes:
+                    self.events_view.selectionModel().select(index,
+                                                             QtCore.QItemSelectionModel.SelectionFlag.Select |
+                                                             QtCore.QItemSelectionModel.SelectionFlag.Rows)
+                self.programmatic_select = False
+
+        if action == 'active_select':
+            self.move_to_trash_action.setEnabled(False)
+            self.restore_from_trash_action.setEnabled(False)
+            self.delete_action.setEnabled(False)
+            self.new_event_action.setEnabled(False)
+            self.export_action.setEnabled(False)
+
+            if uuids:
+                if len(uuids) == 1:
+                    self.new_event_action.setEnabled(True)
+
+                enable_restore = False
+                enable_move_to_trash = False
+                for entity in map(get_entity_from_uuid_safe, uuids):
+                    if entity.is_removed():
+                        enable_restore |= True
+                    else:
+                        enable_move_to_trash |= True
+
+                self.restore_from_trash_action.setEnabled(enable_restore)
+                self.move_to_trash_action.setEnabled(enable_move_to_trash)
+                self.delete_action.setEnabled(True)
+                self.export_action.setEnabled(True)
+
+    def __current_event_changed(self, _: QtCore.QModelIndex, __: QtCore.QModelIndex):
+        if self.programmatic_select:
+            return
+
+        uuids = [index.data(UUIDRole) for index in self.events_view.selectedIndexes() if index.column() == 0]
+        self.state.updated('active_select', tscat._Event, uuids)
+
+    def __catalogue_selection_changed(self, _: QtCore.QItemSelection, __: QtCore.QItemSelection) -> None:
+        if self.programmatic_select:
+            return
+
+        uuids = [index.data(UUIDRole) for index in self.catalogues_view.selectedIndexes()]
+        self.state.updated('active_select', tscat._Catalogue, uuids)
+
+    def __undo_redo_index_changed(self, index: int) -> None:
+        def __create_action_list_menu(index_range: range,
```

Comment:
this could be a module free function

Reply:
You mean, I could put this function somewhere else? It captures and uses `self`.

Reply:
It captures and uses  `self`. What do you mean by module-free?

Reply:
free apply to function, if there is no way to make a function then a method it should be.